src/dotty/tools/dotc/transform/Space.scala

```diff
@@ -0,0 +1,313 @@
+      isSubType(tp1, tp2) ||
+      simplify(a) == Empty ||
+      (isSubType(tp2, tp1) &&
+        canDecompose(tp1) &&
+        subspace(Or(getPartitions(tp1).map(Typ)), b))
+    case (Kon(_, _), Or(_)) =>
+      simplify(minus(a, b)) == Empty
+    case (Kon(tp1, ss1), Kon(tp2, ss2)) =>
+      isEqualType(tp1, tp2) && ss1.zip(ss2).forall((subspace _).tupled)
+  }
+
+  /** intersection of two spaces  */
+  def intersect(a: Space, b: Space): Space = (a, b) match {
+    case (Empty, _) | (_, Empty) => Empty
+    case (_, Or(ss)) => Or(ss.map(intersect(a, _)))
+    case (Or(ss), _) => Or(ss.map(intersect(_, b)))
+    case (Typ(tp1), Typ(tp2)) =>
+      if (isSubType(tp1, tp2)) a
+      else if (isSubType(tp2, tp1)) b
+      else Empty
+    case (Typ(tp1), Kon(tp2, ss)) =>
+      if (isSubType(tp2, tp1)) b
+      else if (isSubType(tp1, tp2)) a
+      else Empty
+    case (Kon(tp1, ss), Typ(tp2)) =>
+      if (isSubType(tp1, tp2) || isSubType(tp2, tp1)) a
+      else Empty
+    case (Kon(tp1, ss1), Kon(tp2, ss2)) =>
+      if (!isEqualType(tp1, tp2)) Empty
+      else if (ss1.zip(ss2).exists(p => simplify(intersect(p._1, p._2)) == Empty)) Empty
+      else
+        Kon(tp1, ss1.zip(ss2).map((intersect _).tupled))
+  }
+
+  /**  the space of a not covered by b */
+  def minus(a: Space, b: Space): Space = (a, b) match {
+    case (Empty, _) => Empty
+    case (_, Empty) => a
+    case (Typ(tp1), Typ(tp2)) =>
+      if (isSubType(tp1, tp2)) Empty
+      else if (isSubType(tp2, tp1) && canDecompose(tp1))
+        minus(Or(getPartitions(tp1).map(Typ)), b)
+      else a
+    case (Typ(tp1), Kon(tp2, ss)) =>
+      if (isSubType(tp1, tp2)) minus(Kon(tp2, getSignature(tp2).map(Typ)), b)
+      else if (isSubType(tp2, tp1) && canDecompose(tp1))
+        minus(Or(getPartitions(tp1).map(Typ)), b)
+      else a
+    case (_, Or(ss)) =>
+      ss.foldLeft(a)(minus)
+    case (Or(ss), _) =>
+      Or(ss.map(minus(_, b)))
+    case (Kon(tp1, ss), Typ(tp2)) =>
+      if (isSubType(tp1, tp2)) Empty
+      else if (simplify(a) == Empty) Empty
+      else if (isSubType(tp2, tp1) && canDecompose(tp1))
+        minus(Or(getPartitions(tp1).map(Typ)), b)
+      else a
+    case (Kon(tp1, ss1), Kon(tp2, ss2)) =>
+      if (!isEqualType(tp1, tp2)) a
+      else if (ss1.zip(ss2).exists(p => simplify(intersect(p._1, p._2)) == Empty)) a
+      else if (ss1.zip(ss2).forall((subspace _).tupled)) Empty
+      else
+        Or(
+          ss1.zip(ss2).map((minus _).tupled).zip(0 to ss2.length - 1).map {
+            case (ri, i) => Kon(tp1, ss1.updated(i, ri))
+          })
+  }
+}
+
+/** scala implementation of space logic */
+class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
+  import tpd._
+
+  def debug(s: String): Unit = {
+    if (ctx.debug) println(s)
+  }
+
+  /** if roundUp, approximate extractors to its type, else approximate extractors to Empty */
+  def project(pat: Tree)(implicit ctx: Context, roundUp: Boolean = true): Space = pat match {
```

Comment:
It seems we do need to pass default parameters around, in case it's a user-supplied value.